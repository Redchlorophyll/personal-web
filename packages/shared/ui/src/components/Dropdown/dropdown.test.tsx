import React, { useState } from 'react';
import { render, screen } from '@testing-library/react';
import '@testing-library/jest-dom';
import { Dropdown } from './index';
import userEvent from '@testing-library/user-event';
import { mockAllIsIntersecting } from 'react-intersection-observer/test-utils';
import { act } from 'react-dom/test-utils';

type optVal = {
  label: string;
  value: string;
};

type props = {
  isCombobox: boolean;
};

function DropdownWrapper(props: props) {
  const [options] = useState<Array<optVal>>([
    { label: 'BCA', value: 'bca' },
    { label: 'BRI', value: 'bri' },
    { label: 'Mandiri', value: 'mandiri' },
    { label: 'SeaBank', value: 'seabank' },
    { label: 'myBank', value: 'myBank' },
  ]);
  const [activeBank, setActiveBank] = useState<string>('');

  return (
    <div>
      <button>trigger blur</button>

      <Dropdown
        type={props.isCombobox ? 'combobox' : 'dropdown'}
        options={options}
        placeholder="pilih bank"
        value={activeBank}
        onChange={(val: optVal | undefined) => {
          if (val) {
            setActiveBank(val.value);
          }
        }}
      />
    </div>
  );
}

describe('shared - ui - dropdown', () => {
  test('it should render properly', () => {
    render(<DropdownWrapper isCombobox={false} />);
    mockAllIsIntersecting(true);

    const target = screen.getByPlaceholderText('pilih bank');
    expect(target).toBeInTheDocument();
  });

  test('it should show right value', async () => {
    const user = userEvent.setup();
    render(<DropdownWrapper isCombobox={false} />);
    mockAllIsIntersecting(true);

    const target = screen.getByPlaceholderText('pilih bank');
    await act(async () => await user.click(target));
    const selectedOpt = screen.getByText('BCA');
    await act(async () => await user.click(selectedOpt));
    expect(screen.getByDisplayValue('BCA')).toBeInTheDocument();
  });

  it('should change combobox value if changed by input', async () => {
    const user = userEvent.setup();
    render(<DropdownWrapper isCombobox={true} />);
    mockAllIsIntersecting(true);

    const target = screen.getByPlaceholderText('pilih bank');
    await act(async () => await user.type(target, 'testing input'));

    expect(screen.queryByDisplayValue('testing input')).toBeInTheDocument();
  });

  it('should show suggestion when input exists in options', async () => {
    const user = userEvent.setup();
    render(<DropdownWrapper isCombobox={true} />);
    mockAllIsIntersecting(true);

    const target = screen.getByPlaceholderText('pilih bank');
    await act(async () => await user.type(target, 'bc'));
    expect(screen.queryByText('BCA')).toBeInTheDocument();
    expect(screen.queryByText('mandiri')).not.toBeInTheDocument();
  });

  it('should clear value selection', async () => {
    const user = userEvent.setup();
    render(<DropdownWrapper isCombobox={false} />);
    mockAllIsIntersecting(true);

    const target = screen.getByPlaceholderText('pilih bank');
    await act(async () => await user.click(target));
    expect(screen.queryByText('Clear Selection')).not.toBeInTheDocument();
    const selectedOpt = screen.getByText('BCA');
    await act(async () => await user.click(selectedOpt));
    expect(screen.getByDisplayValue('BCA')).toBeInTheDocument();

    await act(async () => await user.click(target));

    const clearSelection = screen.getByText('Clear Selection');
    await act(async () => await user.click(clearSelection));
    expect(screen.getByDisplayValue('')).toBeInTheDocument();
    expect(screen.getByPlaceholderText('pilih bank')).toBeInTheDocument();
  });
});
