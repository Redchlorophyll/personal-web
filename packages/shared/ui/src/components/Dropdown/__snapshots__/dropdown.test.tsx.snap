// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`shared - ui - dropdown should render properly 1`] = `
<body>
  <div>
    <div>
      <button>
        trigger blur
      </button>
      <div
        class="flex flex-col items-start"
      >
        <div
          class="w-full relative z-[5] dark:text-black-900"
          data-testid="focus-element"
        >
          <input
            class="peer cursor-pointer bg-black-100 w-full p-[5px_17px_5px_13px] border-black-800 border-solid border-[0.5px] focus:outline-none focus:border-solid focus:border-[0.5px] focus:border-primary-800 rounded-lg dark:focus:drop-shadow-[0px_1px_17px_#406fcb]"
            id=""
            placeholder="pilih bank"
            readonly=""
            type="text"
            value=""
          />
          <svg
            class="z-1 absolute left-full top-[13px] -translate-x-[28px] transition-transform peer-focus:transition-transform peer-focus:rotate-180"
            fill="none"
            height="10"
            width="16"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M14.202 0a2.576 2.576 0 0 1-.047 3.643l-6.36 6.194L5.999 7.99 14.202 0Z"
              fill="url(#ic_arrow_svg__a)"
            />
            <path
              d="M1.839.093 9.635 8.04 7.796 9.844 1.804 3.736A2.576 2.576 0 0 1 1.84.093Z"
              fill="#D9D9D9"
            />
            <defs>
              <lineargradient
                gradientUnits="userSpaceOnUse"
                id="ic_arrow_svg__a"
                x1="12.056"
                x2="6.673"
                y1="4.033"
                y2="9.293"
              >
                <stop
                  stop-color="#CBCBCB"
                />
                <stop
                  offset="1"
                  stop-color="#BCC0CA"
                />
              </lineargradient>
            </defs>
          </svg>
        </div>
      </div>
    </div>
  </div>
</body>
`;
